src/main/ruby/core/string.rb

```diff
@@ -220,15 +220,15 @@ def scan(pattern)
       val = (match.length == 1 ? match[0] : match.captures)
       val.taint if taint
 
-      if block_given?
-        Truffle.invoke_primitive(:regexp_set_last_match, match)
+      if block
+        Truffle::RegexpOperations.set_last_match(match, block.binding)
         yield(val)
       else
         ret << val
       end
     end
 
-    Truffle.invoke_primitive(:regexp_set_last_match, last_match)
+    Truffle::RegexpOperations.set_last_match(last_match, Truffle.invoke_primitive(:caller_binding))
```

Comment:
Should this be the `block`'s binding if there is a block?
Are they always the same? I guess yes if it's a literal block but not necessarily otherwise.
It'd be great to add a spec to verify this as the handling of `$~` is fairly complicated in `scan`.